Drop-In/src/main/java/com/braintreepayments/api/dropin/AddCardActivity.java

```diff
@@ -283,19 +286,26 @@ protected void createCard() {
                     .postalCode(cardForm.getPostalCode())
                     .validate(shouldVault);
 
-            if (shouldRequestThreeDSecureVerification()) {
-                ThreeDSecure.performVerification(mBraintreeFragment, cardBuilder,
-                        mDropInRequest.getAmount());
-            } else {
-                Card.tokenize(mBraintreeFragment, cardBuilder);
-            }
+            Card.tokenize(mBraintreeFragment, cardBuilder);
         }
     }
 
     @Override
     public void onPaymentMethodNonceCreated(PaymentMethodNonce paymentMethod) {
-        mBraintreeFragment.sendAnalyticsEvent("sdk.exit.success");
-        finish(paymentMethod, null);
+        if (shouldRequestThreeDSecureVerification() && mDropInRequest.getThreeDSecureRequest() == null) {
+            ThreeDSecureRequest threeDSecureRequest = new ThreeDSecureRequest().amount(mDropInRequest.getAmount());
+            mDropInRequest.threeDSecureRequest(threeDSecureRequest);
+        }
+
+        if (!mRequestedThreeDSecure && shouldRequestThreeDSecureVerification()) {
```

Comment:
Maybe instead of `mRequestedThreeDSecure` this could be called `mPerformedThreeDSecureVerification`? On a first read-through, I assumed this variable indicated whether the merchant had requested 3DS, but that seems to be covered by `shouldRequestThreeDSecureVerification()`.

Reply:
I also wonder whether `mRequestedThreeDSecure` could be checked _inside_ of `shouldRequestThreeDSecureVerification()`. I think that might be a little cleaner, but I'm not sure how much refactoring that would take since `shouldRequestThreeDSecureVerification()` lives in a base class.